sgp30/sgp30.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2021 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package sgp30
+
+import (
+	"encoding/binary"
+	"errors"
+	"log"
+	"strconv"
+	"time"
+
+	"periph.io/x/conn/v3"
+	"periph.io/x/conn/v3/i2c"
+)
+
+const (
+	InitAirQuality       uint16 = 0x2003
+	MeasureAirQuality    uint16 = 0x2008
+	GetIAQBaseline       uint16 = 0x2015
+	SetIAQBaseline       uint16 = 0x201e
+	SetHumidity          uint16 = 0x2061
+	MeasureTest          uint16 = 0x2032
+	GetFeatureSetVersion uint16 = 0x202f
+	MeasureRawSignals    uint16 = 0x2050
+	GetTVOCBaseline      uint16 = 0x20b3
+	SetTVOCBaseline      uint16 = 0x2077
+
+	I2CAddress uint16 = 0x58
+)
+
+// commandDuration maps the defined maximum measurement duration from the sensor
+var commandDuration = map[uint16]time.Duration{
+	InitAirQuality:       time.Millisecond * 10,
+	MeasureAirQuality:    time.Millisecond * 12,
+	GetIAQBaseline:       time.Millisecond * 10,
+	SetIAQBaseline:       time.Millisecond * 10,
+	SetHumidity:          time.Millisecond * 10,
+	MeasureTest:          time.Millisecond * 220,
+	GetFeatureSetVersion: time.Millisecond * 10,
+	MeasureRawSignals:    time.Millisecond * 25,
+	GetTVOCBaseline:      time.Millisecond * 10,
+	SetTVOCBaseline:      time.Millisecond * 10,
+}
+
+// commandResponseLength maps the defined response length including the CRC
+var commandResponseLength = map[uint16]int{
+	MeasureAirQuality:    6,
+	GetIAQBaseline:       6,
+	MeasureTest:          3,
+	GetFeatureSetVersion: 3,
+	MeasureRawSignals:    6,
+	GetTVOCBaseline:      3,
+}
+
+// CO2 represents the current carbon dioxide value in ppm
+type CO2 uint16
+
+func (c CO2) String() string {
+	return strconv.Itoa(int(c)) + "ppm"
+}
+
+func (c *CO2) Set(b []byte) {
+	*c = (CO2)(binary.BigEndian.Uint16(b))
+}
+
+// TVOC represents the current total volatile organic compounds value in ppb
+type TVOC uint16
+
+func (t TVOC) String() string {
+	return strconv.Itoa(int(t)) + "ppb"
+}
+
+func (t *TVOC) Set(b []byte) {
+	*t = (TVOC)(binary.BigEndian.Uint16(b))
+}
+
+// Env represents measurements from an environmental sensor.
+type Env struct {
+	CO2  CO2
+	TVOC TVOC
+}
+
+// NewI2C returns an object that communicates over I2C to SGP30 environmental sensor.
+//
+// The address must be 0x58.
+func NewI2C(b i2c.Bus, addr uint16) (*Dev, error) {
+	switch addr {
+	case I2CAddress:
+	default:
+		return nil, errors.New("sgp30: given address not supported by device")
+	}
+	d := &Dev{
+		d: &i2c.Dev{Bus: b, Addr: addr},
+		env: Env{
+			CO2:  400,
+			TVOC: 0,
+		},
+	}
+	if err := d.makeDev(); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// Dev is a handle to an initialized SGP30 device.
+type Dev struct {
+	d   conn.Conn
+	env Env
+}
+
+// EquivalentCO2 returns the latest equivalent CO2 value
```

Comment:
Why not expose Env instead of having methods? This could read as the last read value. But this has an inherent race condition, so I recommend against.

Reply:
Ignore.